### DISCUSSION TITLE
Conditional access to parent scope in nested fragment

### DISCUSSION THREAD

**Discussion by ckalan** [alpinejs/alpine#4026] [1. Help] [answered]
What is the best practice for an html fragment containing alpine attributes like x-show that is included in several pages where some pages don't contain an `x-data` element?  If we don't make sure a variable is defined above, we get an undefined variable error. 

Example HTML Fragment:

```
<button x-show="!inModal">Back</button>
<button :disabled="loading" >Save</button>
```

Here, we have a Back button but it is only visible when the containing page is not loaded inside a modal. If it is inside a modal, Back button doesn't make sense so it should be hidden.

Any page that includes this HTML must define `x-data` with `inModal` and `loading` variables or we must use a global store. Otherwise, Alpine throws errors. What we want is to set defaults to such variables in case there is no scope or these variables are not defined.   

We tried this but it doesn't work:

```
<div x-init="loading ||= false; inModal ||= false;">
<button x-show="!inModal">Back</button>
<button :disabled="loading" >Save</button>
</div>
```

This workaround works but it is kind of verbose:

```
<div x-data="{
inModal: typeof(modal) === "undefined" ? false : modal, 
loading: typeof(loading) === "loading" ? false : loading
}">
<button x-show="!inModal">Back</button>
<button :disabled="loading" >Save</button>
</div>
```

Is there a better alternative? And can we make these undefined checks less verbose with a custom magic or sth?

**Comment by SimoTod** [comment, marked as answer]
There's nothing in the core but it's easy enough to define your own magic function. Let's say `$optional`.

```html
<div x-data="{bar: true}">
  <div x-show="$optional('foo')">This will not show but won't trigger an error</div>
  <div x-show="$optional('bar')">This will show</div>
</div>
...
<script>
  document.addEventListener('alpine:init', () => {
    Alpine.magic('optional', (el) => variable => {
      return Alpine.$data(el)[variable] ?? null; 
    })
  })
</script>
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
```

The only suboptimal thing is that you need to use quotes around the variable name inside `$optional` so it gets passed into the magic as a string and it doesn't error.

**Reply by ckalan** [reply]
I was thinking about the same but I couldn't figure out how to check the scope via Alpine.$data and I wasn't aware that I should pass the el to the $data method. Thanks for pointing out the right direction!

**Reply by ckalan** [reply]
I am planning to do sth like this: 
```
<div x-data="$mergeData({ foo: false })" >
<div x-show="foo">..</div>
</div>
```

Using your implementation, I think I can check if foo exists in $data, if not I can add it. I am not sure that newly added property would be reactive though. I'll give it a test.

**Reply by ckalan** [reply]
If it doesn't work, I can use a data instead of magic like 

```
<script>
Alpine.data("mergeData", (value) => {
    return {
      init() {
         // Check each property of value in <this> and add missing ones using Object.assign(this, {...missing})
      }
   } 
}) 
</script>
```

```
<div x-data="mergeData({ foo: false })" >
  <div x-show="foo">..</div>
</div>
```

What do you think about this approach compared to the magic?

**Reply by ckalan** [reply]
And I think your optional would be a very nice candidate for a magic that ships with Alpine. It would be nice to have sth like nullif/coalesce SQL functions with a default value like 

`<div x-show="$ifdef('foo', false)" >`

Don't know if it could handle functions in addition to simple properties

**Reply by ekwoka** [reply]
You can probably just do `x-show="$data.foo"`

**Comment by ekwoka** [comment]
Could your templating not allow you to decide at that level whether you want the alpine directives or not?

**Reply by ckalan** [reply]
It could but then we have to pass state from the client to the server via query/form params and the conditional html rendering would be kind of cumbersome.

**Reply by ekwoka** [reply]
why would you need to pass state? The rendering isn't based on state

And why would it be cumbersome? Doesn't your templating engine already allow conditional rendering?
